Softeq.XToolkit.PushNotifications.Droid/DroidPushNotificationsService.cs

```diff
@@ -0,0 +1,174 @@
+            }
+        }
+
+        protected override void UnregisterFromPushTokenInSystem()
+        {
+            // TODO: possibly use topics and UnsubscribeFromTopic instead
+            Task.Run(() =>
+            {
+                try
+                {
+                    FirebaseInstanceId.Instance.DeleteInstanceId(); //Throws exception if there's no Internet Connection
+                    var token = FirebaseInstanceId.Instance.Token; // Value is null here. This call is needed to force new token generation
+                }
+                catch (Exception e)
+                {
+                    Logger.Warn($"Firebase DeleteInstance failed: {e.Message}");
+                }
+            });
+        }
+
+        public override void ClearAllNotifications()
+        {
+            if (_appContext != null)
+            {
+                var notificationManager = NotificationManager.FromContext(_appContext);
+                notificationManager.CancelAll();
+            }
+        }
+
+        public override PushNotificationModel OnMessageReceived(object pushNotification)
+        {
+            var parsedNotification = base.OnMessageReceived(pushNotification);
+            if (!parsedNotification.IsSilent)
+            {
+                ShowNotification(pushNotification, parsedNotification);
+            }
+            return parsedNotification;
+        }
+
+        private void ShowNotification(object pushNotification, PushNotificationModel parsedPushNotification)
+        {
+            if (_showForegroundNotificationsInSystem || !parsedPushNotification.HandledBySystem)
+            {
+                var notificationData = (pushNotification as RemoteMessage)?.Data;
+                NotificationsHelper.CreateNotification(_appContext, parsedPushNotification, notificationData, _notificationsSettings);
+            }
+        }
+
+        private void OnPushTokenRefreshed(string token)
+        {
+            // To avoid not needed requests only handle this if we were previously registered (have a push token saved) and token changed or if token was not ready at the moment of registration
+            if (_registrationRequired || (!string.IsNullOrEmpty(PushTokenStorageService.PushToken) && PushTokenStorageService.PushToken != token))
+            {
+                _registrationRequired = false;
+                OnRegisteredForPushNotificaions(token);
+            }
+        }
+
+        private void OnNotificationReceived(object pushNotification)
+        {
+            OnMessageReceived(pushNotification);
+        }
+
+        #region IDisposable
+        protected virtual void Dispose(bool disposing)
+        {
+            if (disposing)
+            {
+                XFirebaseIIDService.OnTokenRefreshed -= OnPushTokenRefreshed;
+                XFirebaseMessagingService.OnNotificationReceived -= OnNotificationReceived;
+            }
+        }
+
+        public void Dispose()
```

Comment:
I see `dispose`, but you always have 1 instance of this object(you have to register it as Singleton as I understand). So, the question is, should you call dispose? I think in the real case, `dispose` never been called, so we can safely remove it.

@wcoder what do you think?